File: lt_sfx.c
/*******************************************************
 **             LaserTank ver 4.0                     **
 **               By Jim Kindley                      **
 **               (c) 2001                            **
 **                  Sound Unit                       **
 *******************************************************)

  All Sounds are Wav files loaded into the resource file.
  Because lcc-win32 can't handle user defined resources, the wav files
  are converted into a charecter stream and imported as RCData
*/
#include <windows.h>            // required for all Windows applications
#include <mmsystem.h>
#include "lt_sfx.h"            // prototypes specific to this application

HGLOBAL SFx[MaxSounds];
int LastSFWord	= 0;
int SFXError = 0;
int Sound_On = 1;				// true when sound is on
extern HINSTANCE hInst;			// Globally Defined Instance


// Declare Local Function
static void SoundLoad ( char *s );

void SoundPlay( int sn)
{
char *p;
 if (!Sound_On) return;
 if (SFXError) return;
 p = LockResource(SFx[sn]);
 PlaySound(p,0,5);  // 5 = memory & async
}

static void SoundLoad(char *s)
{
HGLOBAL h;

  if (SFXError) return;
  h = LoadResource(hInst,FindResource(hInst, s, RT_RCDATA));
  if (h == 0)
		SFXError = LastSFWord+1;
  else {
    LastSFWord++;
    SFx[LastSFWord] = h; // put resource handle in Sound Array
  }
}

void SFxInit( void)
{
 SoundLoad("BRICKS"); SoundLoad("FIRE");
 SoundLoad("MOVE"); SoundLoad("HEAD");
 SoundLoad("TURN"); SoundLoad("ENDLEV");
 SoundLoad("DIE"); SoundLoad("ANTI1");
 SoundLoad("ANTI2"); SoundLoad("DEFLB");
 SoundLoad("LASER2"); SoundLoad("PUSH2");
 SoundLoad("PUSH1"); SoundLoad("ROTATE");
 SoundLoad("PUSH3"); SoundLoad("SINK");
 if (SFXError) MessageBox(0,"Loading Error","SFx Error",MB_OK);
}

File: lt_sfx.pas
(*******************************************************
 **             Laser Tank ver 2.0a                   **
 **               By Jim Kindley                      **
 **               (c) 1998,1997,1996                  **
 **                  Sound Unit                       **
 *******************************************************)

 { All Sounds are Wav files loaded into the lt_sfx.res resource file.
   I created a new type of resource in the resource workshop, called "wSound".
   Then you can compile the wave files into the resource
}
unit lt_sfx;
 
interface
{$R lt_sfx.res }
uses mmsystem,winprocs,wintypes;

const
 MaxSounds         = 20; { the maximum number of Sounds }
 LastSFWord : word = 0;
 SFXError: Word    = 0;
 snd_NoStop        = $0010;  { don't stop any currently playing sound }
 ResTypeName       = 'wSound';
 Sound_On: Boolean = true;   { true when Sound is On }

var
 SFx: Array[1..MaxSounds] of THandle;

 Procedure SoundPlay( sn: word);  { SoundPlay(5) Play sound 5 }
 Procedure SFxInit;                      { Load all sounds }
 Procedure SoundKill;                    { This Unloads All Data }

implementation

Procedure SoundPlay( sn: Word);
var p: pointer;
begin
 if Not Sound_On then exit;
 p := LockResource(SFx[sn]);
 sndplaySound(p,5);  { 5 = memory & async }
 UnLockResource(SFx[sn]);
end;

Procedure SoundLoad(S: pChar);
var h: Thandle;
begin
  h := LoadResource(hInstance,FindResource(hInstance,s,ResTypeName));
  if h = 0 then SFXError := LastSFWord+1
  else begin
    inc(LastSFWord);
    SFx[LastSFWord] := h; { put resource handle in Sound Array }
  end;
end;

Procedure SFxInit;
begin
 SoundLoad('Explos');
 SoundLoad('Laser'); SoundLoad('Move');
 SoundLoad('Ping'); SoundLoad('Movet');
 SoundLoad('EndLev'); SoundLoad('Die');
 if SFXError <> 0 then messagebox(0,'Loading Error','Error',mb_ok);
end;

Procedure SoundKill;
          { run this to release all the memory used to store the sounds }
var i: word;
begin
 For i := 1 to LastSFWord do FreeResource(SFx[i]);
 LastSFWord := 0;
end;

end.